File: exp/separate-rpq-results.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <unordered_map>
#include <set>
#include <string>


using namespace std;

#define TIME_OUT 60000000000
#define NQ 21

int main(int argc, char **argv)
{   

    if (argc < 3) {
        cerr << "  Usage: " << argv[0] << " <query-results-file> <query-types-file> <excluded-queries-file>" << endl;
        exit(1);
    }

    std::string line;
    
    uint64_t n_line = 0, q = 0;

    unordered_map<string, uint64_t> M;

    M["v /* c"] = 1;
    M["v * c"] = 2;
    M["v + c"] = 3;
    M["c * v"] = 4;
    M["c /* v"] = 5;
    M["v / c"] = 6;
    M["v / v"] = 7;
    M["v */* c"] = 8;
    M["v |* c"] = 9;
    M["v | v"] = 10;
    M["v ^ v"] = 11;
    M["v /* v"] = 12;
    M["v */*/*/*/* c"] = 13;
    M["v * v"] = 14;
    M["v /? c"] = 15;
    M["v + v"] = 16;
    M["v /+ c"] = 17;
    M["v ^/ c"] = 18;
    M["v || v"] = 19;
    M["v /^ v"] = 20;
    M["v | c"] = 21;

    std::ofstream fp[NQ+3];

    for (int i =  1; i <= NQ+2; i++) {
        fp[i].open(string(argv[1])+"-"+to_string(i), ios::out); 
        fp[i] << "\\begin{filecontents}{q" << i << "-" << argv[1] << ".dat}" << endl;
    }

    std::ofstream fp_all, fp_v_to_v, fp_c_to_v;
    fp_all.open(string(argv[1])+"-all");
    fp_v_to_v.open(string(argv[1])+"-all_v_to_v");
    fp_c_to_v.open(string(argv[1])+"-all_c_to_v");

    std::ifstream ifs_results(argv[1], std::ifstream::in);
    std::ifstream ifs_types(argv[2], std::ifstream::in);
    std::ifstream ifs_excluded(argv[3], std::ifstream::in);

    std::set<uint64_t> m_excluded;
    uint64_t eq;

    ifs_excluded >> eq;
    do {
        m_excluded.insert(eq);
	cout << "Excluding query " << eq << "..." << endl;
    } while (ifs_excluded >> eq);


    int nQ, nResults;
    int64_t qTime;
    char semicolon;

    unordered_map<uint64_t, string> nq_to_qtype;

    for (int i = 1; i <= 2110; i++) {
        getline(ifs_types, line);
        nq_to_qtype[i] = line;
    }


    for (int i = 1; i <= 1589; i++) {
        //getline(ifs_types, line);
	ifs_results >> nQ;
	ifs_results >> semicolon;
	ifs_results >> nResults;
	ifs_results >> semicolon;
	ifs_results >> qTime;

//	if (m_excluded.find(i) != m_excluded.end()) continue;

	if (nResults == -1 || nResults == -2 || qTime > TIME_OUT) qTime = TIME_OUT;

        fp_all << (float)qTime/1000000000 << endl;

        if (nq_to_qtype[nQ].at(0)=='v' and nq_to_qtype[nQ].at(nq_to_qtype[nQ].size()-1)=='v')
            fp_v_to_v << (float)qTime/1000000000 << endl;
	else
            fp_c_to_v << (float)qTime/1000000000 << endl;

        if (M.find(nq_to_qtype[nQ]) != M.end()) {
	    //if (i == 320) cout << line << endl;	
	    fp[M[nq_to_qtype[nQ]]] << (float)qTime/1000000000 << endl;
	}
	else {
	    if (nq_to_qtype[nQ].at(0) == 'c' or nq_to_qtype[nQ].at(nq_to_qtype[nQ].size()-1) == 'c')
	        fp[NQ+1] << (float)qTime/1000000000 << endl;
	    else 
		fp[NQ+2] << (float)qTime/1000000000 << endl;
	}

	//cout << nQ << ";" << nResults << ";" << qTime << endl;
    }


    for (int i =  1; i <= NQ+2; i++) {
        fp[i] << "\\end{filecontents}";
    }

    return 0;
}

File: exp/separate-rpq-results.py
import sys
import os
import math

TIME_OUT = 60000000000
NQ = 21

if len(sys.argv) < 4:
    print(f"Usage: {sys.argv[0]} <query-results-file> <query-types-file> <selected-queries-file>")
    sys.exit(1)

line = ""
n_line = 0
q = 0

M = {
    "v /* c": 1,
    "v * c": 2,
    "v + c": 3,
    "c * v": 4,
    "c /* v": 5,
    "v / c": 6,
    "v / v": 7,
    "v */* c": 8,
    "v |* c": 9,
    "v | v": 10,
    "v ^ v": 11,
    "v /* v": 12,
    "v */*/*/*/* c": 13,
    "v * v": 14,
    "v /? c": 15,
    "v + v": 16,
    "v /+ c": 17,
    "v ^/ c": 18,
    "v || v": 19,
    "v /^ v": 20,
    "v | c": 21
}

fp = [None] + [open(sys.argv[1] + "-" + str(i), "w") for i in range(1, NQ + 3)]
for i in range(1, NQ + 3):
    fp[i].write(f"\\begin{{filecontents}}{{q{i}-{sys.argv[1]}.dat}}\n")

fp_all = open(sys.argv[1] + "-all", "w")
fp_v_to_v = open(sys.argv[1] + "-all_v_to_v", "w")
fp_c_to_v = open(sys.argv[1] + "-all_c_to_v", "w")

m_excluded = set()
with open(sys.argv[3], "r") as ifs_excluded:
    for line in ifs_excluded:
        eq = int(line.strip())
        m_excluded.add(eq)
        print(f"Including query {eq}...")

nq_to_qtype = {}
with open(sys.argv[2], "r") as ifs_types:
    for i in range(1, 2111):
        line = ifs_types.readline().strip()
        nq_to_qtype[i] = line

with open(sys.argv[1], "r") as ifs_results:
    for i in range(1, 2111):
        nQ, semicolon, nResults, semicolon, qTime = map(int, ifs_results.readline().strip().split())

        if i not in m_excluded:
            continue

        if nResults == -1 or nResults == -2 or qTime > TIME_OUT:
            qTime = TIME_OUT

        fp_all.write(f"{qTime / 1000000000:.9f}\n")

        if nq_to_qtype[nQ][0] == "v" and nq_to_qtype[nQ][-1] == "v":
            fp_v_to_v.write(f"{qTime / 1000000000:.9f}\n")
        else:
            fp_c_to_v.write(f"{qTime / 1000000000:.9f}\n")

        if nq_to_qtype[nQ] in M:
            fp[M[nq_to_qtype[nQ]]].write(f"{qTime / 1000000000:.9f}\n")

